frontend/src/lib/UriUtil.test.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * @license
+ * Copyright 2018-2019 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { buildHttpUri, buildWsUri, getWindowBaseUriParts } from "./UriUtil"
+
+const location = {}
+
+global.window = Object.create(window)
+Object.defineProperty(window, "location", { value: location })
+
+test("gets all window URI parts", () => {
+  location.hostname = "the_host"
+  location.port = "9988"
+  location.pathname = "foo"
+
+  const parts = getWindowBaseUriParts()
+  expect(parts).toStrictEqual({
+    host: "the_host",
+    port: 9988,
+    basePath: "foo",
+  })
+})
+
+test("gets window URI parts without basePath", () => {
+  location.hostname = "the_host"
+  location.port = "9988"
+  location.pathname = ""
+
+  const parts = getWindowBaseUriParts()
+  expect(parts).toStrictEqual({
+    host: "the_host",
+    port: 9988,
+    basePath: "",
+  })
+})
+
+test("gets window URI parts with long basePath", () => {
+  location.hostname = "the_host"
+  location.port = "9988"
+  location.pathname = "/foo/bar"
+
+  const parts = getWindowBaseUriParts()
+  expect(parts).toStrictEqual({
+    host: "the_host",
+    port: 9988,
+    basePath: "foo/bar",
+  })
+})
+
+test("gets window URI parts with weird basePath", () => {
+  location.hostname = "the_host"
+  location.port = "9988"
+  location.pathname = "///foo/bar//"
+
+  const parts = getWindowBaseUriParts()
+  expect(parts).toStrictEqual({
+    host: "the_host",
+    port: 9988,
+    basePath: "foo/bar",
+  })
+})
+
+test("builds HTTP URI correctly", () => {
```

Comment:
All tests from here on down are unrelated to this PR, but I thought I might was well implement anyway.